client/client.go

```diff
@@ -151,142 +115,221 @@ func New(url string, args ...interface{}) *Client {
+				request.publishing,
 			)
 
 			if err != nil {
-				logger.Warn("could not publish message")
-				request.errChan <- err
-
-				// The message that we tried to publish is NOT added back to the
-				// queue since it never left the client. The sender will get
-				// an error back and should handle this manually!
-				continue
+				// We failed to publish the message so we don't need the correlation mapping any more
+				// because a new one will be added the next time it's retried. And if it's not retried
+				// we don't need it any more anyway.
+				c.mu.Lock()
+				delete(c.correlationMapping, request.publishing.CorrelationId)
+				c.mu.Unlock()
+
+				if request.numRetries >= 0 {
+					// The message that we tried to publish is NOT added back to the
+					// queue since it never left the client. The sender will get
+					// an error back and should handle this manually!
+					logger.Warn("client: could not publish message, giving up")
+					request.errChan <- err
+				} else {
+					logger.Warn("client: could not publish message, retrying")
+					request.numRetries += 1
+					c.clientMessages <- request
+				}
+
+				return
 			}
 
-			// If no error is met, close the error channel so the sender doesn't
-			// hang forever.
+			// The request has left the client, no more errors can occur from here.
 			close(request.errChan)
+
+			logger.Info("client: did publish message")
 		}
 	}()
+}
 
-	// We consume messages from the same queue as we publish them but we watch for
-	// responses in a separate go routine. All requests sent will get a response
-	// back on the consumer channel but we only pass them to the user of the
-	// client if we mapped the correlation ID to a response channel.
-	go func() {
-		for {
-			response, _ := <-messages
+// runRepliesConsumer will declare and start consuming from the queue where we expect replies to come back.
+// The method will stop consuming if the underlying amqp channel is closed for any reason.
+func (c *Client) runRepliesConsumer(inChan *amqp.Channel) error {
+	queue, err := inChan.QueueDeclare(
+		c.replyToQueueName,
+		c.queueDeclareSettings.Durable,
+		c.queueDeclareSettings.DeleteWhenUnused,
+		c.queueDeclareSettings.Exclusive,
+		c.queueDeclareSettings.NoWait,
+		c.queueDeclareSettings.Args,
+	)
+	if err != nil {
+		return err
+	}
 
-			// Check if the published message correlation ID is maped.
-			// This should be true as long as the caller of publish requested a reply.
-			if replyChannel, ok := queueChannels[response.CorrelationId]; ok {
-				replyChannel <- &response
+	messages, err := inChan.Consume(
+		queue.Name,
+		c.consumeSettings.Consumer,
+		c.consumeSettings.AutoAck,
+		c.consumeSettings.Exclusive,
+		c.consumeSettings.NoLocal,
+		c.consumeSettings.NoWait,
+		c.consumeSettings.Args,
+	)
+
+	if err != nil {
+		return err
+	}
 
-				// Remove the mapping between correlation ID and reply channel.
-				delete(queueChannels, response.CorrelationId)
+	go func() {
```

Comment:
Should be moved to a named function.

Reply:
You mean setting up the amqp channels and starting the consumer in once function and the loop in another? Yea, that sound good.